#!/usr/bin/env nextflow

/*
========================================================================================
                  C H I P - S E Q   B E S T   P R A C T I C E
========================================================================================
 ChIP-seq Best Practice Analysis Pipeline. Started May 2016.
 #### Homepage / Documentation
 https://github.com/nf-core/chipseq
 @#### Authors
 Chuan Wang <chuan.wang@scilifelab.se>
 Phil Ewels <phil.ewels@scilifelab.se>
 Alex Peltzer <alexander.peltzer@qbic.uni-tuebingen.de>
 Harshil Patel <harshil.patel@crick.ac.uk>
----------------------------------------------------------------------------------------
*/

def helpMessage() {
    log.info nfcoreHeader()
    log.info"""
    Usage:

    The typical command for running the pipeline is as follows:

      nextflow run nf-core/chipseq --design design.csv --genome GRCh37 -profile docker

    Mandatory arguments:
      --design                      Comma-separted file containing information about the samples in the experiment (see docs/usage.md)
      --fasta                       Path to Fasta reference. Not mandatory when using reference in iGenomes config via --genome
      --gtf                         Path to GTF file in Ensembl format. Not mandatory when using reference in iGenomes config via --genome
      -profile                      Configuration profile to use. Can use multiple (comma separated)
                                    Available: conda, docker, singularity, awsbatch, test

    Generic
      --genome                      Name of iGenomes reference
      --singleEnd                   Specifies that the input is single-end reads
      --narrowPeak                  Run MACS in narrowPeak mode. Default: broadPeak
      --fragment_size [int]         Estimated fragment size used to extend single-end reads. Default: 0

    References                      If not specified in the configuration file or you wish to overwrite any of the references
      --bwa_index                   Full path to directory containing BWA index including base name i.e. /path/to/index/genome.fa
      --gene_bed                    Path to BED file containing gene intervals
      --tss_bed                     Path to BED file containing transcription start sites (used by ataqv)
      --macs_gsize                  Effective genome size parameter required by MACS2. If using iGenomes config, values have only been provided when --genome is set as GRCh37, GRCm38, hg19, mm10, BDGP6 and WBcel235
      --blacklist                   Path to blacklist regions (.BED format), used for filtering alignments
      --saveGenomeIndex             If generated by the pipeline save the BWA index in the results directory

    Trimming
      --clip_r1 [int]               Instructs Trim Galore to remove bp from the 5' end of read 1 (or single-end reads)
      --clip_r2 [int]               Instructs Trim Galore to remove bp from the 5' end of read 2 (paired-end reads only)
      --three_prime_clip_r1 [int]   Instructs Trim Galore to remove bp from the 3' end of read 1 AFTER adapter/quality trimming has been performed
      --three_prime_clip_r2 [int]   Instructs Trim Galore to re move bp from the 3' end of read 2 AFTER adapter/quality trimming has been performed
      --skipTrimming                Skip the adapter trimming step
      --saveTrimmed                 Save the trimmed FastQ files in the the Results directory

    Alignments
      --keepDups                    Duplicate reads are not filtered from alignments
      --keepMultiMap                Reads mapping to multiple locations are not filtered from alignments
      --saveAlignedIntermediates    Save the intermediate BAM files from the alignment step - not done by default

    Other
      --outdir                      The output directory where the results will be saved
      --email                       Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits
      --maxMultiqcEmailFileSize     Theshold size for MultiQC report to be attached in notification email. If file generated by pipeline exceeds the threshold, it will not be attached (Default: 25MB)
      -name                         Name for the pipeline run. If not specified, Nextflow will automatically generate a random mnemonic

    AWSBatch
      --awsqueue                    The AWSBatch JobQueue that needs to be set when running on AWSBatch
      --awsregion                   The AWS Region for your AWS Batch job to run on
    """.stripIndent()
}

///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
/* --                                                                     -- */
/* --                SET UP CONFIGURATION VARIABLES                       -- */
/* --                                                                     -- */
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////

/*
 * SET UP CONFIGURATION VARIABLES
 */

// Show help message
if (params.help){
    helpMessage()
    exit 0
}

// Check if genome exists in the config file
if (params.genomes && params.genome && !params.genomes.containsKey(params.genome)) {
    exit 1, "The provided genome '${params.genome}' is not available in the iGenomes file. Currently the available genomes are ${params.genomes.keySet().join(", ")}"
}

////////////////////////////////////////////////////
/* --         DEFAULT PARAMETER VALUES         -- */
////////////////////////////////////////////////////

// Configurable variables
params.fasta = params.genome ? params.genomes[ params.genome ].fasta ?: false : false
params.bwa_index = params.genome ? params.genomes[ params.genome ].bwa ?: false : false
params.gtf = params.genome ? params.genomes[ params.genome ].gtf ?: false : false
params.gene_bed = params.genome ? params.genomes[ params.genome ].gene_bed ?: false : false
params.macs_gsize = params.genome ? params.genomes[ params.genome ].macs_gsize ?: false : false
params.blacklist = params.genome ? params.genomes[ params.genome ].blacklist ?: false : false

// Has the run name been specified by the user?
//  this has the bonus effect of catching both -name and --name
custom_runName = params.name
if (!(workflow.runName ==~ /[a-z]+_[a-z]+/)){
  custom_runName = workflow.runName
}

////////////////////////////////////////////////////
/* --          CONFIG FILES                    -- */
////////////////////////////////////////////////////

// Pipeline config
output_docs_ch = Channel.fromPath("$baseDir/docs/output.md", checkIfExists: true)

// JSON files required by BAMTools for alignment filtering
if (params.singleEnd) {
    bamtools_filter_config_ch = Channel.fromPath(params.bamtools_filter_se_config, checkIfExists: true)
} else {
    bamtools_filter_config_ch = Channel.fromPath(params.bamtools_filter_pe_config, checkIfExists: true)
}

// Header files for MultiQC
multiqc_config_ch = Channel.fromPath(params.multiqc_config, checkIfExists: true)
mlib_peak_count_header_ch = Channel.fromPath("$baseDir/assets/multiqc/mlib_peak_count_header.txt", checkIfExists: true)
mlib_frip_score_header_ch = Channel.fromPath("$baseDir/assets/multiqc/mlib_frip_score_header.txt", checkIfExists: true)
mlib_peak_annotation_header_ch = Channel.fromPath("$baseDir/assets/multiqc/mlib_peak_annotation_header.txt", checkIfExists: true)
mlib_deseq2_pca_header_ch = Channel.fromPath("$baseDir/assets/multiqc/mlib_deseq2_pca_header.txt", checkIfExists: true)
mlib_deseq2_clustering_header_ch = Channel.fromPath("$baseDir/assets/multiqc/mlib_deseq2_clustering_header.txt", checkIfExists: true)

////////////////////////////////////////////////////
/* --          VALIDATE INPUTS                 -- */
////////////////////////////////////////////////////

// Validate inputs
if (params.design){
    design_csv = Channel
        .fromPath(params.design, checkIfExists: true)
        .ifEmpty { exit 1, "Samples design file not found: ${params.design}" }
} else {
    exit 1, "Samples design file not specified!"
}

if (params.fasta){
    lastPath = params.fasta.lastIndexOf(File.separator)
    bwa_base = params.fasta.substring(lastPath+1)

    Channel
        .fromPath(params.fasta, checkIfExists: true)
        .ifEmpty { exit 1, "Fasta file not found: ${params.fasta}" }
        .into { fasta_bwa_index;
                fasta_genome_filter;
                fasta_markdup_metrics;
                fasta_mlib_metrics;
                fasta_mlib_macs_annotate;
                fasta_mlib_macs_consensus_annotate;
                fasta_mrep_macs_annotate;
                fasta_mrep_macs_consensus_annotate;
                fasta_igv }
} else {
    exit 1, "Fasta file not specified!"
}

if (params.gtf){
    Channel
        .fromPath(params.gtf, checkIfExists: true)
        .ifEmpty { exit 1, "GTF annotation file not found: ${params.gtf}" }
        .into { gtf_gene_bed;
                gtf_mlib_macs_annotate;
                gtf_mlib_macs_consensus_annotate;
                gtf_mrep_macs_annotate;
                gtf_mrep_macs_consensus_annotate }
} else {
    exit 1, "GTF annotation file not specified!"
}

if (params.bwa_index){
    lastPath = params.bwa_index.lastIndexOf(File.separator)
    bwa_dir =  params.bwa_index.substring(0,lastPath+1)
    bwa_base = params.bwa_index.substring(lastPath+1)

    bwa_index = Channel
        .fromPath(bwa_dir, checkIfExists: true)
        .ifEmpty { exit 1, "BWA index directory not found: ${bwa_dir}" }
}

if (params.gene_bed){
    gene_bed = Channel
        .fromPath(params.gene_bed, checkIfExists: true)
        .ifEmpty { exit 1, "Gene BED annotation file not found: ${params.gene_bed}" }
}

if (params.tss_bed){
    tss_bed = Channel
        .fromPath(params.tss_bed, checkIfExists: true)
        .ifEmpty { exit 1, "TSS BED annotation file not found: ${params.tss_bed}" }
}

if (params.blacklist) {
    blacklist = Channel
        .fromPath(params.blacklist, checkIfExists: true)
        .ifEmpty { exit 1, "Blacklist file not found: ${params.blacklist}" }
}

////////////////////////////////////////////////////
/* --                   AWS                    -- */
////////////////////////////////////////////////////

if( workflow.profile == 'awsbatch') {
  // AWSBatch sanity checking
  if (!params.awsqueue || !params.awsregion) exit 1, "Specify correct --awsqueue and --awsregion parameters on AWSBatch!"
  // Check outdir paths to be S3 buckets if running on AWSBatch
  // related: https://github.com/nextflow-io/nextflow/issues/813
  if (!params.outdir.startsWith('s3:')) exit 1, "Outdir not on S3 - specify S3 Bucket to run on AWSBatch!"
  // Prevent trace files to be stored on S3 since S3 does not support rolling files.
  if (workflow.tracedir.startsWith('s3:')) exit 1, "Specify a local tracedir or run without trace! S3 cannot be used for tracefiles."
}






////////////////////////////////////////////////////
/* --          CONFIG FILES                    -- */
////////////////////////////////////////////////////

// Check for file existence
macsconfig = file(params.macsconfig)
if( !macsconfig.exists() ) exit 1, "Missing MACS config: '$macsconfig'. Specify path with --macsconfig"

if( params.bwa_index ){
    lastPath = params.bwa_index.lastIndexOf(File.separator)
    bwa_dir =  params.bwa_index.substring(0,lastPath+1)
    bwa_base = params.bwa_index.substring(lastPath+1)

    Channel
        .fromPath(bwa_dir, checkIfExists: true)
        .ifEmpty { exit 1, "BWA index directory not found: ${bwa_dir}" }
        .set { bwa_index }
} else if ( params.fasta ){
    lastPath = params.fasta.lastIndexOf(File.separator)
    bwa_base = params.fasta.substring(lastPath+1)

    fasta = Channel
        .fromPath(params.fasta, checkIfExists: true)
        .ifEmpty { exit 1, "Fasta file not found: ${params.fasta}" }
} else {
    exit 1, "No reference genome specified!"
}

if( params.gtf ){
    gtf = Channel
        .fromPath(params.gtf, checkIfExists: true)
        .ifEmpty { exit 1, "GTF annotation file not found: ${params.gtf}" }
}

if( params.bed ){
    bed = Channel
        .fromPath(params.bed, checkIfExists: true)
        .ifEmpty { exit 1, "BED file not found: ${params.bed}" }
}

if( params.blacklist_filtering ){
    blacklist = Channel
        .fromPath(params.blacklist, checkIfExists: true)
        .ifEmpty { exit 1, "Blacklist annotation file not found: ${params.blacklist}" }
}

if(params.readPaths){
    if(params.singleEnd){
        Channel
            .from(params.readPaths)
            .map { row -> [ row[0], [file(row[1][0])]] }
            .ifEmpty { exit 1, "params.readPaths was empty - no input files supplied" }
            .into { raw_reads_configvali; raw_reads_fastqc; raw_reads_trimgalore }
    } else {
        Channel
            .from(params.readPaths)
            .map { row -> [ row[0], [file(row[1][0]), file(row[1][1])]] }
            .ifEmpty { exit 1, "params.readPaths was empty - no input files supplied" }
            .into { raw_reads_configvali; raw_reads_fastqc; raw_reads_trimgalore }
    }
} else {
    Channel
        .fromFilePairs( params.reads, size: params.singleEnd ? 1 : 2 )
        .ifEmpty { exit 1, "Cannot find any reads matching: ${params.reads}\nNB: Path needs to be enclosed in quotes!\nIf this is single-end data, please specify --singleEnd on the command line." }
        .into { raw_reads_configvali; raw_reads_fastqc; raw_reads_trimgalore }
}

/*
 * Create a channel for macs config file
 */
Channel
    .from(macsconfig.readLines())
    .map { line ->
        list = line.split(',')
        def chip_sample_id = list[0]
        def ctrl_sample_id = list[1]
        def analysis_id = list[2]
        [ chip_sample_id, ctrl_sample_id, analysis_id ]
    }
    .into { vali_para; macs_para; saturation_para }

// Validate all samples in macs config file
def config_samples = []
for (line in vali_para){
    if (line.getClass().toString() != "class groovyx.gpars.dataflow.operator.PoisonPill") {
        config_samples.add(line[0])
        config_samples.add(line[1])
    }
}
config_samples.removeAll{ it == '' }
config_samples.unique(false)

def fastq_samples = []
for (sample in raw_reads_configvali){
    if (sample.getClass().toString() != "class groovyx.gpars.dataflow.operator.PoisonPill") {
        fastq_samples.add(sample[0].toString() - ~/(.R)?(_R)?(.R1)?(_1)?(_R1)?(_trimmed)?(_val_1)?(\.fq)?(\.fastq)?(\.gz)?$/)
    }
}

def missing_samples = config_samples - config_samples.intersect(fastq_samples)
if(!missing_samples.isEmpty()){
    exit 1, "No FastQ file found for sample in MACS config: ${missing_samples}"
}

def dropped_samples = fastq_samples - fastq_samples.intersect(config_samples)
if(!dropped_samples.isEmpty()){
    exit 1, "Sample ${dropped_samples} not included in MACS config"
}


log.info nfcoreHeader()
def summary = [:]
summary['Run Name']             = custom_runName ?: workflow.runName
summary['Reads']                = params.reads
summary['Data Type']            = params.singleEnd ? 'Single-End' : 'Paired-End'
summary['Genome']               = params.genome
if(params.bwa_index)  summary['BWA Index'] = params.bwa_index
else if(params.fasta) summary['Fasta File'] = params.fasta
if(params.largeRef)  summary['Build BWA Index for Large Reference'] = params.largeRef
if(params.gtf)  summary['GTF File'] = params.gtf
if(params.bed)  summary['BED File'] = params.bed
summary['Blacklist Filtering']  = params.blacklist_filtering
if(params.blacklist_filtering) summary['Blacklist BED'] = params.blacklist
summary['Save Reference']       = params.saveReference
summary['Multiple Alignments']  = params.allow_multi_align
summary['Duplication Removal']  = params.skipDupRemoval
if(params.seqCenter) summary['Seq Center'] = params.seqCenter
summary['Save Intermeds']       = params.saveAlignedIntermediates
summary['Fingerprint Bins']     = params.fingerprintBins
summary['MACS Config']          = params.macsconfig
summary['MACS Broad Peaks']     = params.broad
summary['MACS Genome Size']     = params.macsgsize
summary['Saturation Analysis']  = params.saturation
summary['Extend Reads']         = "$params.extendReadsLen bp"
if( params.notrim ){
    summary['Trimming Step'] = 'Skipped'
} else {
    summary['Trim R1'] = params.clip_r1
    summary['Trim R2'] = params.clip_r2
    summary["Trim 3' R1"] = params.three_prime_clip_r1
    summary["Trim 3' R2"] = params.three_prime_clip_r2
}
summary['Save Trimmed']         = params.saveTrimmed
summary['Max Resources']        = "$params.max_memory memory, $params.max_cpus cpus, $params.max_time time per job"
if(workflow.containerEngine) summary['Container'] = "$workflow.containerEngine - $workflow.container"
summary['Output Dir']           = params.outdir
summary['Launch Dir']           = workflow.launchDir
summary['Working Dir']          = workflow.workDir
summary['Script Dir']           = workflow.projectDir
summary['User']                 = workflow.userName
if (workflow.profile == 'awsbatch'){
   summary['AWS Region']        = params.awsregion
   summary['AWS Queue']         = params.awsqueue
}
summary['Config Profile']       = workflow.profile
if (params.config_profile_description) summary['Config Description'] = params.config_profile_description
if (params.config_profile_contact)     summary['Config Contact']     = params.config_profile_contact
if (params.config_profile_url)         summary['Config URL']         = params.config_profile_url
if(params.email) {
  summary['E-mail Address']     = params.email
  summary['MultiQC Max Size']   = params.maxMultiqcEmailFileSize
}
log.info summary.collect { k,v -> "${k.padRight(21)}: $v" }.join("\n")
log.info "\033[2m----------------------------------------------------\033[0m"

// Check the hostnames against configured profiles
checkHostname()

if( workflow.profile == 'awsbatch') {
  // AWSBatch sanity checking
  if (!params.awsqueue || !params.awsregion) exit 1, "Specify correct --awsqueue and --awsregion parameters on AWSBatch!"
  // Check outdir paths to be S3 buckets if running on AWSBatch
  // related: https://github.com/nextflow-io/nextflow/issues/813
  if (!params.outdir.startsWith('s3:')) exit 1, "Outdir not on S3 - specify S3 Bucket to run on AWSBatch!"
  // Prevent trace files to be stored on S3 since S3 does not support rolling files.
  if (workflow.tracedir.startsWith('s3:')) exit 1, "Specify a local tracedir or run without trace! S3 cannot be used for tracefiles."
}


/*
 * PREPROCESSING - Build BWA index
 */
if(!params.bwa_index && fasta){
    process makeBWAindex {
        tag "$fasta"
        publishDir path: { params.saveReference ? "${params.outdir}/reference_genome" : params.outdir },
                   saveAs: { params.saveReference ? it : null }, mode: 'copy'

        input:
        file fasta from fasta

        output:
        file "BWAIndex" into bwa_index

        script:
        BWAIndexOption = params.largeRef ? "bwtsw" : 'is'
        """
        bwa index -a $BWAIndexOption $fasta
        mkdir BWAIndex && mv ${fasta}* BWAIndex
        """
    }
}


/*
 * PREPROCESSING - Build BED file
 */
if(!params.bed){
    process makeBED {
        tag "$gtf"
        publishDir path: { params.saveReference ? "${params.outdir}/reference_genome" : params.outdir },
                   saveAs: { params.saveReference ? it : null }, mode: 'copy'

        input:
        file gtf from gtf

        output:
        file "${gtf.baseName}.bed" into bed

        script: // This script is bundled with the pipeline, in nfcore/chipseq/bin/
        """
        gtf2bed $gtf > ${gtf.baseName}.bed
        """
    }
}


/*
 * STEP 1 - FastQC
 */
process fastqc {
    tag "$name"
    publishDir "${params.outdir}/fastqc", mode: 'copy',
        saveAs: {filename -> filename.indexOf(".zip") > 0 ? "zips/$filename" : "$filename"}

    input:
    set val(name), file(reads) from raw_reads_fastqc

    output:
    file '*_fastqc.{zip,html}' into fastqc_results
    file '.command.out' into fastqc_stdout

    script:
    """
    fastqc -q $reads
    """
}


/*
 * STEP 2 - Trim Galore!
 */
if(params.notrim){
    trimmed_reads = read_files_trimming
    trimgalore_results = []
    trimgalore_fastqc_reports = []
} else {
    process trim_galore {
        tag "$name"
        publishDir "${params.outdir}/trim_galore", mode: 'copy',
            saveAs: {filename ->
                if (filename.indexOf("_fastqc") > 0) "FastQC/$filename"
                else if (filename.indexOf("trimming_report.txt") > 0) "logs/$filename"
                else params.saveTrimmed ? filename : null
            }
        label 'process_medium'

        input:
        set val(name), file(reads) from raw_reads_trimgalore

        output:
        file '*.fq.gz' into trimmed_reads
        file '*trimming_report.txt' into trimgalore_results
        file "*_fastqc.{zip,html}" into trimgalore_fastqc_reports

        script:
        c_r1 = params.clip_r1 > 0 ? "--clip_r1 ${params.clip_r1}" : ''
        c_r2 = params.clip_r2 > 0 ? "--clip_r2 ${params.clip_r2}" : ''
        tpc_r1 = params.three_prime_clip_r1 > 0 ? "--three_prime_clip_r1 ${params.three_prime_clip_r1}" : ''
        tpc_r2 = params.three_prime_clip_r2 > 0 ? "--three_prime_clip_r2 ${params.three_prime_clip_r2}" : ''
        if (params.singleEnd) {
            """
            trim_galore --fastqc --gzip $c_r1 $tpc_r1 $reads
            """
        } else {
            """
            trim_galore --paired --fastqc --gzip $c_r1 $c_r2 $tpc_r1 $tpc_r2 $reads
            """
        }
    }
}


/*
 * STEP 3.1 - align with bwa
 */
process bwa {
    tag "$prefix"
    publishDir path: { params.saveAlignedIntermediates ? "${params.outdir}/bwa" : params.outdir }, mode: 'copy',
               saveAs: {filename -> params.saveAlignedIntermediates ? filename : null }
    label 'process_big'

    input:
    file reads from trimmed_reads
    file index from bwa_index.collect()

    output:
    file '*.bam' into bwa_bam

    script:
    prefix = reads[0].toString() - ~/(.R1)?(_1)?(_R1)?(_trimmed)?(_val_1)?(\.fq)?(\.fastq)?(\.gz)?$/
    filtering = params.allow_multi_align ? '' : "| samtools view -b -q 1 -F 4 -F 256"
    seqCenter = params.seqCenter ? "-R '@RG\\tID:${prefix}\\tCN:${params.seqCenter}'" : ''
    """
    bwa mem -M $seqCenter ${index}/${bwa_base} $reads | samtools view -bT $index - $filtering > ${prefix}.bam
    """
}


/*
 * STEP 3.2 - post-alignment processing
 */
process samtools {
    tag "${bam.baseName}"
    publishDir path: "${params.outdir}/bwa", mode: 'copy',
                saveAs: { filename ->
                    if (filename.indexOf(".stats.txt") > 0) "stats/$filename"
                    else params.saveAlignedIntermediates ? filename : null
                }
    label 'process_medium'

    input:
    file bam from bwa_bam

    output:
    set file('*.sorted.bam'), file('*.sorted.bam.bai') into bam_picard, bam_for_mapped
    file '*.sorted.bed' into bed_total
    file '*.stats.txt' into samtools_stats

    script:
    """
    samtools sort $bam -o ${bam.baseName}.sorted.bam
    samtools index ${bam.baseName}.sorted.bam
    bedtools bamtobed -i ${bam.baseName}.sorted.bam | sort -k 1,1 -k 2,2n -k 3,3n -k 6,6 > ${bam.baseName}.sorted.bed
    samtools stats ${bam.baseName}.sorted.bam > ${bam.baseName}.stats.txt
    """
}


/*
 * STEP 3.3 - Statistics about mapped and unmapped reads against ref genome
 */

process bwa_mapped {
    tag "${bam.baseName}"
    publishDir "${params.outdir}/bwa/mapped", mode: 'copy'
    label 'process_medium'

    input:
    set file(bam), file(bai) from bam_for_mapped

    output:
    file "${bam.baseName}.mapped_refgenome.txt" into bwa_mapped

    script:
    """
    samtools idxstats $bam \\
        | awk -v filename="$bam" '{mapped+=\$3; unmapped+=\$4} END {print filename,"\t",mapped,"\t",unmapped}' \\
        > ${bam.baseName}.mapped_refgenome.txt
    """
}


/*
 * STEP 4 Picard
 */
if (params.skipDupRemoval) {
    bam_picard.into {
        bam_dedup_spp;
        bam_dedup_deepTools;
        bam_dedup_macs;
        bam_dedup_saturation
    }
    picard_reports = Channel.from(false)
} else {
    process picard {
        tag "$prefix"
        publishDir "${params.outdir}/picard", mode: 'copy'
        label 'process_medium'

        input:
        set file(bam), file(bai) from bam_picard

        output:
        set file("${prefix}.dedup.sorted.bam"), file("${prefix}.dedup.sorted.bam.bai") into bam_dedup_spp, bam_dedup_deepTools, bam_dedup_macs, bam_dedup_saturation
        file "${prefix}.dedup.sorted.bed" into bed_dedup
        file "${prefix}.picardDupMetrics.txt" into picard_reports

        script:
        prefix = bam.toString() - ~/(\.sorted)?(\.bam)?$/
        if( !task.memory ){
            log.info "[Picard MarkDuplicates] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this."
            avail_mem = 3
        } else {
            avail_mem = task.memory.toGiga()
        }
        """
        picard MarkDuplicates \\
            -Xmx${avail_mem}g \\
            INPUT=$bam \\
            OUTPUT=${prefix}.dedup.bam \\
            ASSUME_SORTED=true \\
            REMOVE_DUPLICATES=true \\
            METRICS_FILE=${prefix}.picardDupMetrics.txt \\
            VALIDATION_STRINGENCY=LENIENT \\
            PROGRAM_RECORD_ID='null'

        samtools sort ${prefix}.dedup.bam -o ${prefix}.dedup.sorted.bam
        samtools index ${prefix}.dedup.sorted.bam
        bedtools bamtobed -i ${prefix}.dedup.sorted.bam | sort -k 1,1 -k 2,2n -k 3,3n -k 6,6 > ${prefix}.dedup.sorted.bed
        """
    }
}


/*
 * STEP 5 Read_count_statistics
 */
process countstat {
    publishDir "${params.outdir}/countstat", mode: 'copy'
    label 'process_long'

    input:
    file bed from params.skipDupRemoval ? bed_total.collect() : bed_total.mix(bed_dedup).collect()

    output:
    file 'read_count_statistics.txt' into countstat_results

    script:
    """
    countstat.pl $bed
    """
}


/*
 * STEP 6.1 Phantompeakqualtools
 */

process phantompeakqualtools {
    tag "$prefix"
    publishDir "${params.outdir}/phantompeakqualtools", mode: 'copy',
                saveAs: {filename -> filename.indexOf(".out") > 0 ? "logs/$filename" : "$filename"}
    label 'process_long'

    input:
    set file(bam), file(bai) from bam_dedup_spp
    file phantompeakqualtools_mqc_header from phantompeakqualtools_mqc_header_ch.collect()

    output:
    file '*.pdf' into spp_plot
    file '*.spp.out' into spp_out, spp_out_mqc
    file '*_mqc.csv' into spp_csv_mqc

    script:
    prefix = bam.toString() - ~/(\.dedup)?(\.sorted)?(\.bam)?$/
    """
    run_spp.r -c="$bam" -savp -savd="${prefix}.spp.Rdata" -out="${prefix}.spp.out"
    processSppRdata.r ${prefix}.spp.Rdata ${prefix}.spp.csv
    cat $phantompeakqualtools_mqc_header ${prefix}.spp.csv > ${prefix}_mqc.csv
    """
}


/*
 * STEP 6.2 Combine and calculate NSC & RSC
 */

process calculateNSCRSC {
    publishDir "${params.outdir}/phantompeakqualtools", mode: 'copy'

    input:
    file spp_out_list from spp_out.collect()

    output:
    file 'cross_correlation_processed.txt' into calculateNSCRSC_results

    script:
    """
    cat $spp_out_list > cross_correlation.txt
    calculateNSCRSC.r cross_correlation.txt
    """
}


bam_dedup_deepTools.into {
    bam_dedup_deepTools_bamPEFragmentSize;
    bam_dedup_deepTools_plotFingerprint;
    bam_dedup_deepTools_bamCoverage;
    bam_dedup_deepTools_multiBamSummary
}

/*
 * STEP 7.1 deepTools bamPEFragmentSize
 */
process deepTools_bamPEFragmentSize {
    tag "$bam_base"
    publishDir "${params.outdir}/deepTools/FragmentSize", mode: 'copy'
    label 'process_big'

    input:
    set file(bam), file(bai) from bam_dedup_deepTools_bamPEFragmentSize

    output:
    file '*.{txt,pdf}' into deepTools_bamPEFragmentSize_results
    file '*.txt' into deepTools_bamPEFragmentSize_multiqc

    when:
    !params.singleEnd

    script:
    bam_base = bam.baseName - '.dedup.sorted'
    """
    bamPEFragmentSize \\
        --binSize 1000 \\
        --bamfiles $bam \\
        --o ${bam_base}.fragment_size_histogram.pdf \\
        --numberOfProcessors ${task.cpus} \\
        --plotFileFormat pdf \\
        --plotTitle "${bam_base}: Paired-end Fragment Size Distribution" \\
        --outRawFragmentLengths ${bam_base}.bamPEFragmentSize_rawdata.txt
    """
}

/*
 * STEP 7.2 deepTools plotFingerprint
 */
process deepTools_plotFingerprint {
    publishDir "${params.outdir}/deepTools", mode: 'copy'
    label 'process_big'

    input:
    file(bambai) from bam_dedup_deepTools_plotFingerprint.collect()

    output:
    file '*.{txt,pdf}' into deepTools_plotFingerprint_results
    file '*.txt' into deepTools_plotFingerprint_multiqc

    script:
    """
    plotFingerprint \\
        -b *.bam \\
        --plotFile fingerprint.pdf \\
        --outRawCounts fingerprint.txt \\
        --extendReads ${params.extendReadsLen} \\
        --skipZeros \\
        --ignoreDuplicates \\
        --numberOfSamples ${params.fingerprintBins} \\
        --binSize 500 \\
        --numberOfProcessors ${task.cpus} \\
        --plotFileFormat pdf \\
        --plotTitle "Fingerprint Plot"
    """
}

/*
 * STEP 7.3 deepTools bamCoverage
 */
process deepTools_bamCoverage {
    tag "${bam.baseName - '.dedup.sorted'}"
    publishDir "${params.outdir}/deepTools/bigWig", mode: 'copy'
    label 'process_big'

    input:
    set file(bam), file(bai) from bam_dedup_deepTools_bamCoverage

    output:
    file "${bam.baseName}.bw" into deepTools_bamCoverage_results

    script:
    """
    bamCoverage \\
       -b $bam \\
       --extendReads ${params.extendReadsLen} \\
       --normalizeUsing RPKM \\
       --numberOfProcessors ${task.cpus} \\
       -o ${bam.baseName}.bw
    """
}

/*
 * STEP 7.4 deepTools computeMatrix
 */
process deepTools_computeMatrix {
    publishDir "${params.outdir}/deepTools", mode: 'copy'
    label 'process_big'

    input:
    file bigwig from deepTools_bamCoverage_results.collect()
    file bed from bed.collect()

    output:
    file 'computeMatrix.out.gz' into deepTools_computeMatrix_results

    script:
    """
    computeMatrix scale-regions \\
        --scoreFileName *.bw \\
        --regionsFileName $bed \\
        --beforeRegionStartLength 3000 \\
        --afterRegionStartLength 3000 \\
        --regionBodyLength 5000 \\
        --outFileName computeMatrix.out.gz \\
        --numberOfProcessors ${task.cpus} \\
        --skipZeros \\
        --smartLabels
    """
}


/*
 * STEP 7.5 deepTools plotProfile
 */
process deepTools_plotProfile {
    publishDir "${params.outdir}/deepTools", mode: 'copy'
    label 'process_big'

    input:
    file bigwig from deepTools_computeMatrix_results

    output:
    file '*.{pdf,txt}' into deepTools_plotProfile_results
    file '*.txt' into deepTools_plotProfile_multiqc

    script:
    """
    plotProfile \\
        --matrixFile computeMatrix.out.gz \\
        --outFileName read_distribution_profile.pdf \\
        --plotFileFormat pdf \\
        --outFileNameData read_distribution_profile.txt \\
        --plotTitle "Reads Distribution Profile"
    """
}


/*
 * STEP 7.6 deepTools multiBamSummary
 */
process deepTools_multiBamSummary {
    publishDir "${params.outdir}/deepTools", mode: 'copy'
    label 'process_big'

    input:
    file(bambai) from bam_dedup_deepTools_multiBamSummary.collect()

    output:
    file 'multiBamSummary.npz' into deepTools_multiBamSummary_results_corr, deepTools_multiBamSummary_results_pca

    when:
    bambai.size() > 2

    script:
    """
    multiBamSummary \\
        bins \\
        --binSize 10000 \\
        --bamfiles *.bam \\
        -out multiBamSummary.npz \\
        --extendReads ${params.extendReadsLen} \\
        --ignoreDuplicates \\
        --centerReads \\
        --numberOfProcessors ${task.cpus} \\
        --smartLabels
    """
}


/*
 * STEP 7.7 deepTools plotCorrelation
 */
process deepTools_plotCorrelation {
    publishDir "${params.outdir}/deepTools", mode: 'copy'
    label 'process_big'

    input:
    file npz from deepTools_multiBamSummary_results_corr

    output:
    file '*.{pdf,txt}' into deepTools_plotCorrelation_results
    file '*.txt' into deepTools_plotCorrelation_multiqc

    when:
    !(bam instanceof Path)

    script:
    """
    plotCorrelation \\
        -in $npz \\
        -o heatmap_SpearmanCorr.pdf \\
        --plotFileFormat pdf \\
        --outFileCorMatrix heatmap_SpearmanCorr.txt \\
        --corMethod spearman \\
        --skipZeros \\
        --plotTitle "Spearman Correlation of Read Counts" \\
        --whatToPlot heatmap \\
        --colorMap RdYlBu \\
        --plotNumbers
    """
}


/*
 * STEP 7.8 deepTools plotPCA
 */
process deepTools_plotPCA {
    publishDir "${params.outdir}/deepTools", mode: 'copy'
    label 'process_big'

    input:
    file npz from deepTools_multiBamSummary_results_pca

    output:
    file '*.{pdf,txt}' into deepTools_plotPCA_results
    file '*.txt' into deepTools_plotPCA_multiqc

    when:
    !(bam instanceof Path)

    script:
    """
    plotPCA \\
        -in $npz \\
        -o pcaplot.pdf \\
        --plotFileFormat pdf \\
        --plotTitle "Principal Component Analysis Plot" \\
        --outFileNameData pcaplot.txt \\
        --plotWidth 8
    """
}


/*
 * STEP 8.1 MACS
 */

process macs {
    tag "${analysis_id}"
    publishDir "${params.outdir}/macs", mode: 'copy'
    label 'process_medium'

    input:
    file bam from bam_dedup_macs.collect()
    set chip_sample_id, ctrl_sample_id, analysis_id from macs_para

    output:
    file '*.{bed,r,narrowPeak}' into macs_results
    file '*.xls' into macs_peaks

    script:
    if(!params.skipDupRemoval){
        chip = "-t ${chip_sample_id}.dedup.sorted.bam"
        ctrl = ctrl_sample_id == '' ? '' : "-c ${ctrl_sample_id}.dedup.sorted.bam"
    }
    else {
        chip = "-t ${chip_sample_id}.sorted.bam"
        ctrl = ctrl_sample_id == '' ? '' : "-c ${ctrl_sample_id}.sorted.bam"
    }
    broad = params.broad ? "--broad" : ''
    """
    macs2 callpeak \\
        $chip \\
        $ctrl \\
        $broad \\
        -f BAM \\
        -g $params.macsgsize \\
        -n $analysis_id \\
        -q 0.01
    """
}


/*
 * STEP 8.2 Saturation analysis
 */
if (params.saturation) {

  process saturation {
     tag "${analysis_id}.${sampling}"
     publishDir "${params.outdir}/macs/saturation", mode: 'copy'
     label 'process_medium'

     input:
     file bam from bam_dedup_saturation.collect()
     set chip_sample_id, ctrl_sample_id, analysis_id from saturation_para
     each sampling from 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0

     output:
     file '*.xls' into saturation_results

     script:
     if(!params.skipDupRemoval){
         chip_sample = "${chip_sample_id}.dedup.sorted.bam"
         ctrl = ctrl_sample_id == '' ? '' : "-c ${ctrl_sample_id}.dedup.sorted.bam"
     }
     else {
         chip_sample = "${chip_sample_id}.sorted.bam"
         ctrl = ctrl_sample_id == '' ? '' : "-c ${ctrl_sample_id}.sorted.bam"
     }
     broad = params.broad ? "--broad" : ''
     """
     samtools view -b -s ${sampling} ${chip_sample} > ${chip_sample}.${sampling}.bam
     macs2 callpeak \\
         -t ${chip_sample}.${sampling}.bam \\
         $ctrl \\
         $broad \\
         -f BAM \\
         -g $params.macsgsize \\
         -n ${analysis_id}.${sampling} \\
         -q 0.01
     """
  }

  process saturation_r {
     tag "${saturation_results_collection[0].baseName}"
     publishDir "${params.outdir}/macs/saturation", mode: 'copy'

     input:
     file macsconfig from macsconfig
     file countstat from countstat_results
     file saturation_results_collection from saturation_results.collect()

     output:
     file '*.{txt,pdf}' into saturation_summary

     script:
     """
     saturation_results_processing.r $macsconfig $countstat $saturation_results_collection
     """
  }
}


/*
 * STEP 9 Post peak calling processing
 */

process chippeakanno {
    tag "${macs_peaks_collection[0].baseName}"
    publishDir "${params.outdir}/macs/chippeakanno", mode: 'copy'

    input:
    file macs_peaks_collection from macs_peaks.collect()
    file gtf from gtf

    output:
    file '*.{txt,bed}' into chippeakanno_results

    script:
    filtering = params.blacklist_filtering ? "${params.blacklist}" : "No-filtering"
    """
    post_peak_calling_processing.r $filtering $gtf $macs_peaks_collection
    """
}

/*
 * Parse software version numbers
 */
process get_software_versions {
    publishDir "${params.outdir}/pipeline_info", mode: 'copy',
      saveAs: {filename ->
          if (filename.indexOf(".csv") > 0) filename
          else null
      }

    output:
    file 'software_versions_mqc.yaml' into software_versions_yaml
    file "software_versions.csv"

    script:
    """
    echo $workflow.manifest.version > v_pipeline.txt
    echo $workflow.nextflow.version > v_nextflow.txt
    fastqc --version > v_fastqc.txt || true
    trim_galore --version > v_trim_galore.txt || true
    echo \$(bwa 2>&1) > v_bwa.txt || true
    samtools --version > v_samtools.txt || true
    bedtools --version > v_bedtools.txt || true
    picard MarkDuplicates --version &> v_picard.txt  || true
    echo \$(plotFingerprint --version 2>&1) > v_deeptools.txt || true
    echo \$(macs2 --version 2>&1) > v_macs2.txt || true
    multiqc --version > v_multiqc.txt || true
    scrape_software_versions.py &> software_versions_mqc.yaml
    """
}


/*
 * STEP 10 MultiQC
 */

process multiqc {
    publishDir "${params.outdir}/MultiQC", mode: 'copy'

    input:
    file multiqc_config from multiqc_config_ch.collect().ifEmpty([])
    file ('fastqc/*') from fastqc_results.collect().ifEmpty([])
    file ('trimgalore/*') from trimgalore_results.collect().ifEmpty([])
    file ('samtools/*') from samtools_stats.collect().ifEmpty([])
    file ('picard/*') from picard_reports.collect().ifEmpty([])
    file ('deeptools/bamPEFragmentSize/*') from deepTools_bamPEFragmentSize_multiqc.collect().ifEmpty([])
    file ('deeptools/plotFingerprint/*') from deepTools_plotFingerprint_multiqc.collect().ifEmpty([])
    file ('deeptools/plotProfile/*') from deepTools_plotProfile_multiqc.collect().ifEmpty([])
    file ('deeptools/plotCorrelation/*') from deepTools_plotCorrelation_multiqc.collect().ifEmpty([])
    file ('deeptools/plotPCA/*') from deepTools_plotPCA_multiqc.collect().ifEmpty([])
    file ('phantompeakqualtools/*') from spp_out_mqc.collect().ifEmpty([])
    file ('phantompeakqualtools/*') from spp_csv_mqc.collect().ifEmpty([])
    file ('software_versions/*') from software_versions_yaml.collect().ifEmpty([])

    output:
    file '*multiqc_report.html' into multiqc_report
    file '*_data' into multiqc_data
    file "multiqc_plots"

    script:
    rtitle = custom_runName ? "--title \"$custom_runName\"" : ''
    rfilename = custom_runName ? "--filename " + custom_runName.replaceAll('\\W','_').replaceAll('_+','_') + "_multiqc_report" : ''
    """
    multiqc . -f $rtitle $rfilename --config $multiqc_config \\
        -m custom_content -m fastqc -m cutadapt -m samtools -m picard -m deeptools -m phantompeakqualtools
    """
}

/*
 * STEP 11 - Output Description HTML
 */
process output_documentation {
    publishDir "${params.outdir}/pipeline_info", mode: 'copy'

    input:
    file output_docs from output_docs_ch

    output:
    file "results_description.html"

    script:
    """
    markdown_to_html.r $output_docs results_description.html
    """
}

/*
 * Completion e-mail notification
 */
workflow.onComplete {

    // Set up the e-mail variables
    def subject = "[nf-core/chipseq] Successful: $workflow.runName"
    if(!workflow.success){
      subject = "[nf-core/chipseq] FAILED: $workflow.runName"
    }
    def email_fields = [:]
    email_fields['version'] = workflow.manifest.version
    email_fields['runName'] = custom_runName ?: workflow.runName
    email_fields['success'] = workflow.success
    email_fields['dateComplete'] = workflow.complete
    email_fields['duration'] = workflow.duration
    email_fields['exitStatus'] = workflow.exitStatus
    email_fields['errorMessage'] = (workflow.errorMessage ?: 'None')
    email_fields['errorReport'] = (workflow.errorReport ?: 'None')
    email_fields['commandLine'] = workflow.commandLine
    email_fields['projectDir'] = workflow.projectDir
    email_fields['summary'] = summary
    email_fields['summary']['Date Started'] = workflow.start
    email_fields['summary']['Date Completed'] = workflow.complete
    email_fields['summary']['Pipeline script file path'] = workflow.scriptFile
    email_fields['summary']['Pipeline script hash ID'] = workflow.scriptId
    if(workflow.repository) email_fields['summary']['Pipeline repository Git URL'] = workflow.repository
    if(workflow.commitId) email_fields['summary']['Pipeline repository Git Commit'] = workflow.commitId
    if(workflow.revision) email_fields['summary']['Pipeline Git branch/tag'] = workflow.revision
    if(workflow.container) email_fields['summary']['Docker image'] = workflow.container
    email_fields['skipped_poor_alignment'] = skipped_poor_alignment
    email_fields['summary']['Nextflow Version'] = workflow.nextflow.version
    email_fields['summary']['Nextflow Build'] = workflow.nextflow.build
    email_fields['summary']['Nextflow Compile Timestamp'] = workflow.nextflow.timestamp

    // On success try attach the multiqc report
    def mqc_report = null
    try {
        if (workflow.success) {
            mqc_report = multiqc_report.getVal()
            if (mqc_report.getClass() == ArrayList){
                log.warn "[nf-core/chipseq] Found multiple reports from process 'multiqc', will use only one"
                mqc_report = mqc_report[0]
            }
        }
    } catch (all) {
        log.warn "[nf-core/chipseq] Could not attach MultiQC report to summary email"
    }

    // Render the TXT template
    def engine = new groovy.text.GStringTemplateEngine()
    def tf = new File("$baseDir/assets/email_template.txt")
    def txt_template = engine.createTemplate(tf).make(email_fields)
    def email_txt = txt_template.toString()

    // Render the HTML template
    def hf = new File("$baseDir/assets/email_template.html")
    def html_template = engine.createTemplate(hf).make(email_fields)
    def email_html = html_template.toString()

    // Render the sendmail template
    def smail_fields = [ email: params.email, subject: subject, email_txt: email_txt, email_html: email_html, baseDir: "$baseDir", mqcFile: mqc_report, mqcMaxSize: params.maxMultiqcEmailFileSize.toBytes() ]
    def sf = new File("$baseDir/assets/sendmail_template.txt")
    def sendmail_template = engine.createTemplate(sf).make(smail_fields)
    def sendmail_html = sendmail_template.toString()

    // Send the HTML e-mail
    if (params.email) {
        try {
          if( params.plaintext_email ){ throw GroovyException('Send plaintext e-mail, not HTML') }
          // Try to send HTML e-mail using sendmail
          [ 'sendmail', '-t' ].execute() << sendmail_html
          log.info "[nf-core/chipseq] Sent summary e-mail to $params.email (sendmail)"
        } catch (all) {
          // Catch failures and try with plaintext
          [ 'mail', '-s', subject, params.email ].execute() << email_txt
          log.info "[nf-core/chipseq] Sent summary e-mail to $params.email (mail)"
        }
    }

    // Write summary e-mail HTML to a file
    def output_d = new File( "${params.outdir}/pipeline_info/" )
    if( !output_d.exists() ) {
        output_d.mkdirs()
    }
    def output_hf = new File( output_d, "pipeline_report.html" )
    output_hf.withWriter { w -> w << email_html }
    def output_tf = new File( output_d, "pipeline_report.txt" )
    output_tf.withWriter { w -> w << email_txt }

    c_reset = params.monochrome_logs ? '' : "\033[0m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_red = params.monochrome_logs ? '' : "\033[0;31m";
    if(workflow.success){
        log.info "${c_purple}[nf-core/chipseq]${c_green} Pipeline completed successfully${c_reset}"
    } else {
        checkHostname()
        log.info "${c_purple}[nf-core/chipseq]${c_red} Pipeline completed with errors${c_reset}"
    }

}

def nfcoreHeader(){
    // Log colors ANSI codes
    c_reset = params.monochrome_logs ? '' : "\033[0m";
    c_dim = params.monochrome_logs ? '' : "\033[2m";
    c_black = params.monochrome_logs ? '' : "\033[0;30m";
    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_yellow = params.monochrome_logs ? '' : "\033[0;33m";
    c_blue = params.monochrome_logs ? '' : "\033[0;34m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_cyan = params.monochrome_logs ? '' : "\033[0;36m";
    c_white = params.monochrome_logs ? '' : "\033[0;37m";

    return """    ${c_dim}----------------------------------------------------${c_reset}
                                            ${c_green},--.${c_black}/${c_green},-.${c_reset}
    ${c_blue}        ___     __   __   __   ___     ${c_green}/,-._.--~\'${c_reset}
    ${c_blue}  |\\ | |__  __ /  ` /  \\ |__) |__         ${c_yellow}}  {${c_reset}
    ${c_blue}  | \\| |       \\__, \\__/ |  \\ |___     ${c_green}\\`-._,-`-,${c_reset}
                                            ${c_green}`._,._,\'${c_reset}
    ${c_purple}  nf-core/chipseq v${workflow.manifest.version}${c_reset}
    ${c_dim}----------------------------------------------------${c_reset}
    """.stripIndent()
}

def checkHostname(){
    def c_reset = params.monochrome_logs ? '' : "\033[0m"
    def c_white = params.monochrome_logs ? '' : "\033[0;37m"
    def c_red = params.monochrome_logs ? '' : "\033[1;91m"
    def c_yellow_bold = params.monochrome_logs ? '' : "\033[1;93m"
    if(params.hostnames){
        def hostname = "hostname".execute().text.trim()
        params.hostnames.each { prof, hnames ->
            hnames.each { hname ->
                if(hostname.contains(hname) && !workflow.profile.contains(prof)){
                    log.error "====================================================\n" +
                            "  ${c_red}WARNING!${c_reset} You are running with `-profile $workflow.profile`\n" +
                            "  but your machine hostname is ${c_white}'$hostname'${c_reset}\n" +
                            "  ${c_yellow_bold}It's highly recommended that you use `-profile $prof${c_reset}`\n" +
                            "============================================================"
                }
            }
        }
    }
}
